### PR TITLE
shm15_c kyrtaq

### PR DESCRIPTION
**Pull request by kyrtaq**

### CODE REVIEW

File: shm/ship.cpp [hunk 1]
```diff
@@ -33,3 +33,14 @@ Ship& Ship::operator-=(const size_t crew) {
...
+void Ship::unload(Cargo* cargo) {
+    if (cargo->getAmount()) {
+        auto it = std::find_if(cargo_.begin(), cargo_.end(), [cargo](const std::shared_ptr<Cargo> ptr) {
+            return ptr.get() == cargo;
+        });
+        if (it != cargo_.end()) {
+            cargo_.erase(it);
+        }
+    }
+}
```
Comment (NG90) [inline shm/ship.cpp:46]: This function should find cargo on ship and substract amount, given in Cargo pointer in argument from cargo on ship. Only if whole cargo is going to be unloaded, the cargo should be erased from vector on ship.
In other words, it should be possible to unload not whole amount of requested cargo.

File: shm/ship.cpp [hunk 2]
```diff
@@ -33,3 +33,14 @@ Ship& Ship::operator-=(const size_t crew) {
     crew_ -= crew;
     return *this;
 }
+
+void Ship::unload(Cargo* cargo) {
+    if (cargo->getAmount()) {
```
Comment (NG90) [inline shm/ship.cpp:38]: I do not think this check is needed. If requested amount is equal "0", the cargo will not be simply unloaded.

File: shm/ship.hpp [hunk 3]
```diff
@@ -25,6 +25,9 @@ class Ship {
     Cargo* getCargo(size_t index) { return cargo_[index].get(); }
     std::vector<std::shared_ptr<Cargo>> getCargos() const { return cargo_; }
 
+    void load(std::shared_ptr<Cargo> cargo) { cargo_.push_back(cargo); };
```
Comment (NG90) [inline shm/ship.hpp:28]: This function should check whether requested cargo in not already stored on ship (e.g. some fruit or alcohol) and if it is, except for creating new element in vector, the amount of given cargo should be properly increased.

File: shm/ship.cpp (hunk 2 above)
Comment (kyrtaq) [reply shm/ship.cpp:38]: @NG90 Mateusz specified it in a video

File: shm/ship.cpp [hunk 4]
```diff
@@ -33,3 +33,42 @@ Ship& Ship::operator-=(const size_t crew) {
...
+    });
+    if (it == cargo_.end()) {
+        std::cerr << "yerr we couldnt find this cargo comrade!\n";
+        return;
+    }
+    if ((*it)->getAmount() < cargo->getAmount()) {
+        std::cerr << "cargo would be negative\n";
+        return;
+    }
+    (*it)->operator-=(cargo->getAmount());
```
Comment (NG90) [inline shm/ship.cpp:49]: check if something like this would work:
`*(*it) -= cargo->getAmount()`

File: shm/ship.cpp [hunk 5]
```diff
@@ -33,3 +33,42 @@ Ship& Ship::operator-=(const size_t crew) {
...
+
+void Ship::load(std::shared_ptr<Cargo> cargo) {
+    for (const auto& el : cargo_) {
+        if (el->getName() == cargo->getName()) {
+            *el += cargo->getAmount();
+            return;
+        }
+    }
+    cargo_.emplace_back(cargo);
+}
```
Comment (NG90) [inline shm/ship.cpp]: I think this function should be removed.

File: shm/ship.cpp [hunk 6]
```diff
@@ -33,3 +33,42 @@ Ship& Ship::operator-=(const size_t crew) {
...
+    }
+}
+
+void Ship::load(std::shared_ptr<Cargo> cargo) {
+    for (const auto& el : cargo_) {
+        if (el == cargo) {
+            *el += cargo->getAmount();
+            return;
+        }
+    }
```
Comment (NG90) [inline shm/ship.cpp:62]: nice to have - could be implemented using find_if algorithm.

File: shm/ship.cpp (hunk 4 above)
Comment (kyrtaq) [reply shm/ship.cpp:49]: does work :+1:

File: shm/ship.cpp (hunk 6 above)
Comment (kyrtaq) [reply shm/ship.cpp:62]: if i would do it with a std::find_if i would still have to implement some loop so it doesnt really matter i think

**Pull request merged**